incubator/virtualcluster/pkg/controller/virtualcluster/virtualcluster_controller.go

```diff
@@ -235,30 +295,49 @@ func (r *ReconcileVirtualcluster) createPKISecrets(caGroup *vcpki.ClusterCAGroup
 	return nil
 }
 
+// cleanupVirtualcluster delete all resources related to the given vc
+func (r *ReconcileVirtualcluster) cleanupVirtualcluster(vc *tenancyv1alpha1.Virtualcluster, cv *tenancyv1alpha1.ClusterVersion) {
+	log.Info("NOT IMPLEMENT YET")
+}
+
+// createVirtualcluster creates a new Virtualcluster based on the specified ClusterVersion
 func (r *ReconcileVirtualcluster) createVirtualcluster(vc *tenancyv1alpha1.Virtualcluster, cv *tenancyv1alpha1.ClusterVersion) (reconcile.Result, error) {
 	// 1. create ns
-	if err := r.createNamespace(vc.Name); err != nil {
-		return reconcile.Result{}, err
+	err := r.createNamespace(vc.Name)
+	if err != nil {
+		goto Cleanup
```

Comment:
we could use `defer` to do the cleanup job.
take https://github.com/alibaba/pouch/blob/master/ctrd/container.go#L692 for reference.